Lib/subprocess.py

```diff
@@ -1910,7 +1933,15 @@ def send_signal(self, sig):
             """Send a signal to the process."""
             # Skip signalling a process that we know has already died.
             if self.returncode is None:
-                os.kill(self.pid, sig)
+                if self.__signal_process_group:
```

Comment:
Should I also undo the addition of this magic when shell=True process group killing logic?

The bug was about a hang after the timeout should've expired.  That hang will no longer happen and we will return.  But this does potentially leave the shell's child processes running on some posix platforms and child process configurations.  Likely not what the caller ever intended.

People running on stable Python releases who need to ensure the process group's children have died already need to do their own post-timeout getpgid() + killpg() logic dance similar to what is in here no matter what.  _This_ logic feels like a feature that I'd do for 3.8+ only.

Adding explicit process group killing features to the kill APIs as some of the dup'ed issues have desired is also squarely a feature request, likely a good idea, but belongs in its own set of PRs as it won't be backported as a bugfix.

Reply:
okay, i've convinced myself.  process group related logic belongs in its own PR.  people need to manually specify start_new_session for this logic to be meaningful regardless.  and for uniformity there's a question of what makes sense on windows as it has its own notion of process groups but *not* a start_new_session concept.